File: home/src/spl/home/framing.py
# SPDX-License-Identifier: AGPL-3.0-only
# Copyright (c) 2026 sol pbc

"""Multiplex framing per ../../../proto/framing.md.

Wire format (9-byte header + payload):

    +------+------+---+-------+
    | sid4 | flg1 | len3      |  header
    +------+------+-----------+
    | payload (len bytes)     |
    +-------------------------+

This module implements encoding, decoding, the flag bitfield, the reset
reason codes, and a small `Stream` helper that tracks per-stream credit +
close state per the spec. Protocol violations (reserved bits, illegal flag
combos) raise `ProtocolError`; the caller is expected to translate that
into a RESET frame with reason PROTOCOL_ERROR.

The relay does not parse frames — this code runs on both tunnel endpoints.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Final

# Flag bits — each frame must carry exactly one of OPEN / DATA / CLOSE /
# RESET / WINDOW, except OPEN|DATA (open with initial bytes) and DATA|CLOSE
# (last data + half-close).
FLAG_OPEN: Final[int] = 0x01
FLAG_DATA: Final[int] = 0x02
FLAG_CLOSE: Final[int] = 0x04
FLAG_RESET: Final[int] = 0x08
FLAG_WINDOW: Final[int] = 0x10
FLAG_RESERVED_MASK: Final[int] = 0xE0  # bits 5-7 must be zero on send

# Reset reason codes — 1-byte payload of RESET frames.
RESET_PROTOCOL_ERROR: Final[int] = 0x01
RESET_FLOW_CONTROL_ERROR: Final[int] = 0x02
RESET_STREAM_LIMIT_EXCEEDED: Final[int] = 0x03
RESET_INTERNAL_ERROR: Final[int] = 0x04
RESET_CANCEL: Final[int] = 0x05
RESET_UNSPECIFIED: Final[int] = 0xFF

# Spec constants.
#
# Note: proto/framing.md's prose says "9-byte header" but the field table
# lists 4 + 1 + 3 = 8 bytes (stream_id u32, flags u8, length u24), and the
# ASCII diagram shows two 32-bit rows = 8 bytes. The field table is the
# implementable contract, so 8 it is. A one-line correction to the prose
# is a documentation follow-up.
HEADER_LEN: Final[int] = 8
MAX_PAYLOAD: Final[int] = (1 << 24) - 1  # 16 MiB - 1
INITIAL_WINDOW: Final[int] = 1 << 20  # 1 MiB
MAX_CONCURRENT_STREAMS: Final[int] = 256
RECOMMENDED_CHUNK: Final[int] = 64 * 1024


class ProtocolError(ValueError):
    """Raised when a frame violates proto/framing.md — caller should RESET."""


@dataclass(frozen=True)
class Frame:
    stream_id: int
    flags: int
    payload: bytes

    def encode(self) -> bytes:
        if not 0 <= self.stream_id <= 0xFFFFFFFF:
            raise ProtocolError(f"stream_id out of range: {self.stream_id}")
        if not 0 <= self.flags <= 0xFF:
            raise ProtocolError(f"flags out of range: {self.flags}")
        if self.flags & FLAG_RESERVED_MASK:
            raise ProtocolError(f"reserved flag bits set: {self.flags:#x}")
        length = len(self.payload)
        if length > MAX_PAYLOAD:
            raise ProtocolError(f"payload exceeds 16 MiB - 1: {length}")
        header = bytearray(HEADER_LEN)
        header[0:4] = self.stream_id.to_bytes(4, "big")
        header[4] = self.flags
        header[5:8] = length.to_bytes(3, "big")
        return bytes(header) + self.payload


@dataclass
class FrameDecoder:
    """Stream decoder. Feed bytes, pull frames until no complete frame remains.

    WebSocket message boundaries are ignored — we re-frame ourselves so that
    coalescing or fragmentation at the transport layer doesn't affect
    framing-layer semantics.
    """

    _buf: bytearray = field(default_factory=bytearray)

    def feed(self, data: bytes | bytearray | memoryview) -> None:
        self._buf.extend(data)

    def next(self) -> Frame | None:
        if len(self._buf) < HEADER_LEN:
            return None
        stream_id = int.from_bytes(self._buf[0:4], "big")
        flags = self._buf[4]
        length = int.from_bytes(self._buf[5:8], "big")
        if flags & FLAG_RESERVED_MASK:
            raise ProtocolError(f"reserved flag bits set: {flags:#x}")
        end = HEADER_LEN + length
        if len(self._buf) < end:
            return None
        payload = bytes(self._buf[HEADER_LEN:end])
        del self._buf[:end]
        return Frame(stream_id=stream_id, flags=flags, payload=payload)

    def drain(self) -> list[Frame]:
        out: list[Frame] = []
        while True:
            frame = self.next()
            if frame is None:
                return out
            out.append(frame)


def build_open(stream_id: int, payload: bytes = b"") -> Frame:
    flags = FLAG_OPEN
    if payload:
        flags |= FLAG_DATA
    return Frame(stream_id=stream_id, flags=flags, payload=payload)


def build_data(stream_id: int, payload: bytes, *, close: bool = False) -> Frame:
    if not payload and not close:
        # Empty DATA is legal per spec; preserve caller intent.
        return Frame(stream_id=stream_id, flags=FLAG_DATA, payload=b"")
    flags = FLAG_DATA
    if close:
        flags |= FLAG_CLOSE
    return Frame(stream_id=stream_id, flags=flags, payload=payload)


def build_close(stream_id: int, payload: bytes = b"") -> Frame:
    return Frame(stream_id=stream_id, flags=FLAG_CLOSE, payload=payload)


def build_reset(stream_id: int, reason: int = RESET_UNSPECIFIED) -> Frame:
    if not 0 <= reason <= 0xFF:
        raise ProtocolError(f"reset reason out of range: {reason}")
    return Frame(stream_id=stream_id, flags=FLAG_RESET, payload=bytes([reason]))


def build_window(stream_id: int, credit: int) -> Frame:
    if not 0 <= credit <= 0xFFFFFFFF:
        raise ProtocolError(f"window credit out of range: {credit}")
    return Frame(stream_id=stream_id, flags=FLAG_WINDOW, payload=credit.to_bytes(4, "big"))


def validate_flags(flags: int) -> None:
    """Enforce the valid-combinations rule from proto/framing.md §flags."""
    if flags & FLAG_RESERVED_MASK:
        raise ProtocolError(f"reserved flag bits set: {flags:#x}")
    exclusive = flags & (FLAG_OPEN | FLAG_DATA | FLAG_CLOSE | FLAG_RESET | FLAG_WINDOW)
    if not exclusive:
        raise ProtocolError("frame has no OPEN/DATA/CLOSE/RESET/WINDOW bit set")
    # Allowed: OPEN alone, DATA alone, CLOSE alone, RESET alone, WINDOW alone,
    # OPEN|DATA, DATA|CLOSE. Everything else is a violation.
    allowed = {
        FLAG_OPEN,
        FLAG_DATA,
        FLAG_CLOSE,
        FLAG_RESET,
        FLAG_WINDOW,
        FLAG_OPEN | FLAG_DATA,
        FLAG_DATA | FLAG_CLOSE,
    }
    if exclusive not in allowed:
        raise ProtocolError(f"illegal flag combination: {flags:#x}")


def parse_window_credit(frame: Frame) -> int:
    if frame.flags & (FLAG_OPEN | FLAG_DATA | FLAG_CLOSE | FLAG_RESET) or not (
        frame.flags & FLAG_WINDOW
    ):
        raise ProtocolError("not a WINDOW frame")
    if len(frame.payload) != 4:
        raise ProtocolError(f"WINDOW payload must be 4 bytes, got {len(frame.payload)}")
    return int.from_bytes(frame.payload, "big")


def parse_reset_reason(frame: Frame) -> int:
    if not (frame.flags & FLAG_RESET):
        raise ProtocolError("not a RESET frame")
    if len(frame.payload) != 1:
        raise ProtocolError(f"RESET payload must be 1 byte, got {len(frame.payload)}")
    return frame.payload[0]

File: mobile/src/framing.ts
// SPDX-License-Identifier: AGPL-3.0-only
// Copyright (c) 2026 sol pbc

// Multiplex framing — TS mirror of spl/home/src/spl/home/framing.py.
// Wire format is specified in ../../proto/framing.md; both endpoints must
// stay in lockstep. Changes here need a matching edit on the home side.
//
// Header layout (8 bytes):
//   stream_id : u32 big-endian
//   flags     : u8
//   length    : u24 big-endian

export const FLAG_OPEN = 0x01;
export const FLAG_DATA = 0x02;
export const FLAG_CLOSE = 0x04;
export const FLAG_RESET = 0x08;
export const FLAG_WINDOW = 0x10;
export const FLAG_RESERVED_MASK = 0xe0;

export const RESET_PROTOCOL_ERROR = 0x01;
export const RESET_FLOW_CONTROL_ERROR = 0x02;
export const RESET_STREAM_LIMIT_EXCEEDED = 0x03;
export const RESET_INTERNAL_ERROR = 0x04;
export const RESET_CANCEL = 0x05;
export const RESET_UNSPECIFIED = 0xff;

export const HEADER_LEN = 8;
export const MAX_PAYLOAD = (1 << 24) - 1;
export const INITIAL_WINDOW = 1 << 20; // 1 MiB
export const MAX_CONCURRENT_STREAMS = 256;
export const RECOMMENDED_CHUNK = 64 * 1024;

export class ProtocolError extends Error {
	constructor(message: string) {
		super(message);
		this.name = "ProtocolError";
	}
}

export interface Frame {
	stream_id: number;
	flags: number;
	payload: Uint8Array;
}

export function encodeFrame(frame: Frame): Uint8Array {
	if (frame.stream_id < 0 || frame.stream_id > 0xffffffff) {
		throw new ProtocolError(`stream_id out of range: ${frame.stream_id}`);
	}
	if (frame.flags < 0 || frame.flags > 0xff) {
		throw new ProtocolError(`flags out of range: ${frame.flags}`);
	}
	if (frame.flags & FLAG_RESERVED_MASK) {
		throw new ProtocolError(`reserved flag bits set: ${frame.flags.toString(16)}`);
	}
	const length = frame.payload.byteLength;
	if (length > MAX_PAYLOAD) {
		throw new ProtocolError(`payload exceeds 16 MiB - 1: ${length}`);
	}
	const header = new Uint8Array(HEADER_LEN);
	const view = new DataView(header.buffer);
	view.setUint32(0, frame.stream_id, false);
	view.setUint8(4, frame.flags);
	view.setUint8(5, (length >> 16) & 0xff);
	view.setUint8(6, (length >> 8) & 0xff);
	view.setUint8(7, length & 0xff);
	const out = new Uint8Array(HEADER_LEN + length);
	out.set(header, 0);
	out.set(frame.payload, HEADER_LEN);
	return out;
}

export class FrameDecoder {
	private buf: Uint8Array = new Uint8Array(0);

	feed(chunk: Uint8Array): void {
		if (chunk.byteLength === 0) return;
		const merged = new Uint8Array(this.buf.byteLength + chunk.byteLength);
		merged.set(this.buf, 0);
		merged.set(chunk, this.buf.byteLength);
		this.buf = merged;
	}

	next(): Frame | null {
		if (this.buf.byteLength < HEADER_LEN) return null;
		const view = new DataView(this.buf.buffer, this.buf.byteOffset, this.buf.byteLength);
		const stream_id = view.getUint32(0, false);
		const flags = view.getUint8(4);
		const length = (view.getUint8(5) << 16) | (view.getUint8(6) << 8) | view.getUint8(7);
		if (flags & FLAG_RESERVED_MASK) {
			throw new ProtocolError(`reserved flag bits set: ${flags.toString(16)}`);
		}
		const end = HEADER_LEN + length;
		if (this.buf.byteLength < end) return null;
		const payload = this.buf.slice(HEADER_LEN, end);
		this.buf = this.buf.slice(end);
		return { stream_id, flags, payload };
	}

	drain(): Frame[] {
		const out: Frame[] = [];
		while (true) {
			const frame = this.next();
			if (!frame) return out;
			out.push(frame);
		}
	}
}

export function buildOpen(stream_id: number, payload: Uint8Array = new Uint8Array(0)): Frame {
	return {
		stream_id,
		flags: payload.byteLength ? FLAG_OPEN | FLAG_DATA : FLAG_OPEN,
		payload,
	};
}

export function buildData(
	stream_id: number,
	payload: Uint8Array,
	opts?: { close?: boolean },
): Frame {
	const close = opts?.close ?? false;
	return {
		stream_id,
		flags: close ? FLAG_DATA | FLAG_CLOSE : FLAG_DATA,
		payload,
	};
}

export function buildClose(stream_id: number, payload: Uint8Array = new Uint8Array(0)): Frame {
	return { stream_id, flags: FLAG_CLOSE, payload };
}

export function buildReset(stream_id: number, reason: number = RESET_UNSPECIFIED): Frame {
	if (reason < 0 || reason > 0xff) {
		throw new ProtocolError(`reset reason out of range: ${reason}`);
	}
	return { stream_id, flags: FLAG_RESET, payload: new Uint8Array([reason]) };
}

export function buildWindow(stream_id: number, credit: number): Frame {
	if (credit < 0 || credit > 0xffffffff) {
		throw new ProtocolError(`window credit out of range: ${credit}`);
	}
	const payload = new Uint8Array(4);
	const view = new DataView(payload.buffer);
	view.setUint32(0, credit, false);
	return { stream_id, flags: FLAG_WINDOW, payload };
}

export function parseWindowCredit(frame: Frame): number {
	if (
		!(frame.flags & FLAG_WINDOW) ||
		frame.flags & (FLAG_OPEN | FLAG_DATA | FLAG_CLOSE | FLAG_RESET)
	) {
		throw new ProtocolError("not a WINDOW frame");
	}
	if (frame.payload.byteLength !== 4) {
		throw new ProtocolError(`WINDOW payload must be 4 bytes, got ${frame.payload.byteLength}`);
	}
	const view = new DataView(frame.payload.buffer, frame.payload.byteOffset, 4);
	return view.getUint32(0, false);
}

export function parseResetReason(frame: Frame): number {
	if (!(frame.flags & FLAG_RESET)) {
		throw new ProtocolError("not a RESET frame");
	}
	if (frame.payload.byteLength !== 1) {
		throw new ProtocolError(`RESET payload must be 1 byte, got ${frame.payload.byteLength}`);
	}
	return frame.payload[0];
}

export function validateFlags(flags: number): void {
	if (flags & FLAG_RESERVED_MASK) {
		throw new ProtocolError(`reserved flag bits set: ${flags.toString(16)}`);
	}
	const exclusive = flags & (FLAG_OPEN | FLAG_DATA | FLAG_CLOSE | FLAG_RESET | FLAG_WINDOW);
	if (!exclusive) {
		throw new ProtocolError("frame has no OPEN/DATA/CLOSE/RESET/WINDOW bit set");
	}
	const allowed = new Set([
		FLAG_OPEN,
		FLAG_DATA,
		FLAG_CLOSE,
		FLAG_RESET,
		FLAG_WINDOW,
		FLAG_OPEN | FLAG_DATA,
		FLAG_DATA | FLAG_CLOSE,
	]);
	if (!allowed.has(exclusive)) {
		throw new ProtocolError(`illegal flag combination: ${flags.toString(16)}`);
	}
}
